src/SvmUtils.ts

```diff
@@ -108,19 +109,53 @@ export function getEvent(events: any[], program: PublicKey, eventName: string) {
   throw new Error("Event " + eventName + " not found");
 }
 
+export interface EventType {
+  program: PublicKey;
+  data: any;
+  name: string;
+  slot: number;
+  confirmationStatus: string;
+  blockTime: number;
+  signature: string;
+}
+
 export async function readProgramEvents(
   connection: Connection,
   program: Program<any>,
-  options?: SignaturesForAddressOptions,
-  finality: Finality = "confirmed"
-) {
-  const events = [];
-  const pastSignatures = await connection.getSignaturesForAddress(program.programId, options, finality);
+  finality: Finality = "confirmed",
+  options: SignaturesForAddressOptions = { limit: 1000 }
+): Promise<EventType[]> {
+  const allSignatures: ConfirmedSignatureInfo[] = [];
+
+  // Fetch all signatures in sequential batches
+  while (true) {
+    const signatures = await connection.getSignaturesForAddress(program.programId, options, finality);
+    allSignatures.push(...signatures);
+
+    // Update options for the next batch. Set before to the last fetched signature.
+    if (signatures.length > 0) {
+      options = { ...options, before: signatures[signatures.length - 1].signature };
+    }
 
-  for (const signature of pastSignatures) {
-    events.push(...(await readEvents(connection, signature.signature, [program], finality)));
+    if (options.limit && signatures.length < options.limit) break; // Exit early if the number of signatures < limit
   }
-  return events;
+
+  // Fetch events for all signatures in parallel
+  const eventsWithSlots = await Promise.all(
```

Comment:
This could be resource intensive if there are many signatures. Should we consider batching the Promise.all calls with a configurable batch size to improve efficiency?

Reply:
same net efficiency, right? it's the same number of requests. the bigger issue could be the RPC provider blocking this or having an issue with the number of requests but for now we can leave it like this and improve when we get there.

Reply:
Yeah, I meant splitting in chunks and run them sequentially. Similarly to how we do it here:
https://github.com/UMAprotocol/protocol/blob/e526e5a1caf68ae7be706f6ef9c996660429282a/packages/common/src/EventUtils.ts#L36-L38

But it's ok for now I think!